### PR TITLE
Full address history with DataTables ajax loading

### PR DESCRIPTION
**Pull request by M-Shahbaz**
1. MongoDB collections updated 
2. Storing all address transactions history
3. Limiting display upto `txcount`
4. Address all history using DataTables ajax loading

### CODE REVIEW

**Comment by uaktags** [comment]
Currently testing this out (reindexing now, but also running it on my modified code base), there's a few things that seem to be mid-edit though. I'll try to comment them in the mean time.

File: routes/index.js [hunk 1]
```diff
@@ -196,22 +178,22 @@ router.get('/network', function(req, res) {
 
 router.get('/reward', function(req, res){
   //db.get_stats(settings.coin, function (stats) {
```
Comment (uaktags) [inline routes/index.js]: not sure why this was ever commented out. It's from prior to your PR, so its unrelated, but i noticed it due to the spacial edits.
```suggestion
  db.get_stats(settings.coin, function (stats) {
```

**Comment by uaktags** [comment]
thus far works as expected.
```
ext/getaddresstxsajax?address=<address>&length=50
ext/getaddresstxsajax?address=<address>
```

Shouldn't there be some sort of maximum value somewhere to prevent a dos attack. For instance, my tested address is currently at  
```
"recordsTotal": 8642,
  "recordsFiltered": 8642
```
And im still reindexing, it'd be too simple to hit this guy without a length set and just keep running this query.

**Comment by M-Shahbaz** [comment]
> thus far works as expected.
> 
> ```
> ext/getaddresstxsajax?address=<address>&length=50
> ext/getaddresstxsajax?address=<address>
> ```
> Shouldn't there be some sort of maximum value somewhere to prevent a dos attack. For instance, my tested address is currently at
> 
> ```
> "recordsTotal": 8642,
>   "recordsFiltered": 8642
> ```
> And im still reindexing, it'd be too simple to hit this guy without a length set and just keep running this query.

Yes already prevented with `txcount` limit from settings.json, like we did to display [transactions](https://github.com/iquidus/explorer/issues/256#issuecomment-451921420):

```
app.use('/ext/getaddresstxsajax', function(req,res){
    if(req.query.length > settings.txcount){
        req.query.length = settings.txcount;
    }
```

**Comment by uaktags** [comment]
Correct me if i'm wrong though, but

ext/getaddresstxsajax?address=<address> in this case has no req.query.length specified, so it's null. So it's going to return the full result rather than the settings.txcount.

**Comment by uaktags** [comment]
Yea, so indeed if address is not specified then the txcount isn't used.

perhaps the recommendation would be to fix both of these functions:

```
if(typeof req.query.length === "undefined"){
    req.query.length = settings.txcount;
  } else if(req.query.length > settings.txcount){
      req.query.length = settings.txcount;
  }
```
or maybe simply
```
if(typeof req.query.length === 'undefined' || req.query.length > settings.txcount){
      req.query.length = settings.txcount;
  }
```


```
app.use('/ext/getaddresstxsajax', function(req,res){
  if(typeof req.query.length === 'undefined' || req.query.length > settings.txcount){
    req.query.length = settings.txcount;
  }
```
and
```
app.use('/ext/getlasttxsajax', function(req,res){
  if(typeof req.query.length === 'undefined' || req.query.length > settings.index.last_txs){
    req.query.length = settings.index.last_txs;
  }
```
Now the admin has full control over what they want to allow. Though i think we may want/need a separate setting for this as this is more of a view limit, whereas I believe the original use for txcount was to be a backend setting for how much we want to store in the db. We may want to store more than we want to display such as storing 100k transactions, but only showing 1k (with pagination to allow the rest).

**Comment by M-Shahbaz** [comment]
Updated to fix another vulnerability if the query length is type string:
```
app.use('/ext/getaddresstxsajax', function(req,res){
  req.query.length = parseInt(req.query.length);
  if(isNaN(req.query.length) || req.query.length > settings.index.last_txs){
    req.query.length = settings.txcount;
  }
```
and
```

app.use('/ext/getlasttxsajax', function(req,res){
  req.query.length = parseInt(req.query.length);
  if(isNaN(req.query.length) || req.query.length > settings.index.last_txs){
    req.query.length = settings.index.last_txs;
  }

```

Or as you suggest a more better approach.

**Comment by uaktags** [comment]
I like it, adds a level of customization while mitigating the server-side being overloaded.

**Comment by TheHolyRoger** [comment]
Included in PR https://github.com/iquidus/explorer/pull/319

**Pull request closed without merging**